File: externals/Async-CTP/samples/VB/Windows/Phone/Concert/Finder/Concert/Finder/Providers/EventfulProvider.vb
Imports System
Imports System.Collections.Generic
Imports System.Globalization
Imports System.Linq
Imports System.Net
Imports System.Threading
Imports System.Threading.Tasks
Imports System.Xml.Linq
Imports Concert_Finder.My.Resources
Imports Concert_Finder.ConcertFinder.Model

Namespace ConcertFinder.Providers

    ''' <summary>
    ''' Eventful event data provider.
    ''' </summary>
    Public Class EventfulProvider
        Implements IProvider

        ''' <summary>
        ''' Search for artists and venues based on the given query.
        ''' </summary>
        ''' <param name="query">The search query.</param>
        ''' <param name="cancellationToken">The cancellation token.</param>
        Public Function SearchAsync(query As String, cancellationToken As CancellationToken) As Task(Of SearchResults()) Implements IProvider.SearchAsync
            Dim artistsUri = New Uri([String].Format("http://api.eventful.com/rest/performers/search?keywords={0}&page_size=100&app_key={1}", query, Configuration.EventfulAppID))
            Dim venuesUri = New Uri([String].Format("http://api.eventful.com/rest/venues/search?keywords={0}&page_size=100&app_key={1}", query, Configuration.EventfulAppID))
            Dim artists = DownloadAndParse(artistsUri, query, GetType(Model.Artist), cancellationToken)
            Dim venues = DownloadAndParse(venuesUri, query, GetType(Model.Venue), cancellationToken)
            Return TaskEx.WhenAll(artists, venues)
        End Function

        ''' <summary>Downloads and parses the provided URI.</summary>
        ''' <param name="uri">The URI to download for results.</param>
        ''' <param name="query">The search query.</param>
        ''' <param name="searchType">The search model type.</param>
        ''' <param name="cancellationToken">The cancellation token.</param>
        Private Async Function DownloadAndParse(
             uri As Uri, query As String, searchType As Type, cancellationToken As CancellationToken) As Task(Of SearchResults)
            Try
                Dim content = Await New WebClient().DownloadStringTaskAsync(uri, cancellationToken)
                Dim results = ParseSearchResults(searchType, XElement.Parse(content))
                Return New SearchResults(results, searchType, query)
            Catch
                Return New SearchResults(New List(Of ISearchable)(), searchType, query)
            End Try
        End Function

        ''' <summary>
        ''' Returns a list of events for the artist with the given id.
        ''' </summary>
        ''' <param name="Id">The artist id.</param>
        Public Async Function GetArtistEventsAsync(Id As String) As Task(Of EventsAvailable) Implements IProvider.GetArtistEventsAsync
            Dim Uri = New Uri([String].Format("http://api.eventful.com/rest/events/search?keywords=spid:{0}&page_size=100&category=music&sort_order=date&app_key={1}", Id, Configuration.EventfulAppID))
            Try
                Dim parsed = XElement.Parse(Await New WebClient().DownloadStringTaskAsync(Uri))
                Dim events = From e In parsed.<events>.<event>
                             Let _City = e.<city_name>.Value
                             Let _Region = e.<region_name>.Value
                             Let _Country = e.<country_name>.Value
                             Let _StartTime = If(Not [String].IsNullOrEmpty(e.<start_time>.Value), CDate(e.<start_time>.Value), Nothing)
                             Let _EndTime = If(Not [String].IsNullOrEmpty(e.<stop_time>.Value), CDate(e.<stop_time>.Value), Nothing)
                             Where Not (_StartTime.AddDays(1).CompareTo(_EndTime) < 0)
                             Select New Model.[Event]() With
                                    {
                                        .Id = e.@id,
                                        .Name = e.<title>.Value,
                                        .Location = [String].Format("{0}, {1}", If(_Country.ToLower().Equals("united states"), _Region, _City), _Country),
                                        .Venue = New Venue() With
                                                 {
                                                     .Id = e.<venue_id>.Value,
                                                     .Name = e.<venue_name>.Value,
                                                     .Address = New Address() With
                                                                {
                                                                    .Street = e.<venue_address>.Value,
                                                                    .City = e.<city_name>.Value,
                                                                    .Region = e.Element("region_name").Value,
                                                                    .Country = e.<country_name>.Value,
                                                                    .Latitude = If(Not [String].IsNullOrEmpty(e.<latitude>.Value), CDbl(e.<latitude>.Value), Nothing),
                                                                    .Longitude = If(Not [String].IsNullOrEmpty(e.<longitude>.Value), CDbl(e.Element("longitude").Value), Nothing)
                                                                },
                                                     .Uri = If(Not [String].IsNullOrEmpty(e.<venue_url>.Value), New Uri(e.<venue_url>.Value), Nothing)
                                                },
                                        .StartTime = _StartTime,
                                        .EndTime = _EndTime,
                                        .IsSingleDay = _EndTime - _StartTime <= TimeSpan.FromDays(1),
                                        .IsTimeSpecified = e.<all_day>.Value.Equals("0"),
                                        .Artists = (From a In e.<performers>.<performer>
                                                    Select New Artist() With
                                                           {
                                                               .Id = a.<id>.Value,
                                                               .Name = a.<name>.Value,
                                                               .Uri = New Uri(a.<url>.Value)
                                                           }).ToList(),
                                         .Description = e.<description>.Value,
                                         .Image = If(Not [String].IsNullOrEmpty(e.<image>.<url>.Value) AndAlso Not e.<image>.<url>.Value.EndsWith("gif"), New Uri(e.<image>.<url>.Value), Nothing),
                                         .Uri = New Uri(e.<url>.Value)
                                    }

                Return New EventsAvailable(events.ToList(), EventGroup.Artist, Id)
            Catch ex As Exception
                Return New EventsAvailable(New List(Of Model.Event)(), EventGroup.Artist, Id)
            End Try
        End Function

        ''' <summary>
        ''' Returns a list of events for the venue with the given id.
        ''' </summary>
        ''' <param name="Id">The venue id.</param>
        Public Async Function GetVenueEventsAsync(Id As String) As Task(Of EventsAvailable) Implements IProvider.GetVenueEventsAsync
            Dim Uri = New Uri([String].Format("http://api.eventful.com/rest/events/search?location={0}&page_size=100&category=music&sort_order=date&app_key={1}", Id, Configuration.EventfulAppID))
            Try
                Dim parsed = XElement.Parse(Await New WebClient().DownloadStringTaskAsync(Uri))
                Dim events = From e In parsed.<events>.<event>
                             Let _City = e.<city_name>.Value
                             Let _Region = e.<region_name>.Value
                             Let _Country = e.<country_name>.Value
                             Let _StartTime = If(Not [String].IsNullOrEmpty(e.<start_time>.Value), CDate(e.<start_time>.Value), Nothing)
                             Let _EndTime = If(Not [String].IsNullOrEmpty(e.<stop_time>.Value), CDate(e.<stop_time>.Value), Nothing)
                             Where Not (_StartTime.AddDays(1).CompareTo(_EndTime) < 0)
                             Select New Model.[Event]() With
                                    {
                                        .Id = e.@id,
                                        .Name = e.<title>.Value,
                                        .Location = e.<venue_name>.Value,
                                        .Venue = New Venue() With
                                                 {
                                                     .Id = e.<venue_id>.Value,
                                                     .Name = e.<venue_name>.Value,
                                                     .Address = New Address() With
                                                                {
                                                                    .Street = e.<venue_address>.Value,
                                                                    .City = e.<city_name>.Value,
                                                                    .Region = e.Element("region_name").Value,
                                                                    .Country = e.<country_name>.Value,
                                                                    .Latitude = If(Not [String].IsNullOrEmpty(e.<latitude>.Value), CDbl(e.<latitude>.Value), Nothing),
                                                                    .Longitude = If(Not [String].IsNullOrEmpty(e.<longitude>.Value), CDbl(e.Element("longitude").Value), Nothing)
                                                                },
                                                     .Uri = If(Not [String].IsNullOrEmpty(e.<venue_url>.Value), New Uri(e.<venue_url>.Value), Nothing)
                                                },
                                        .StartTime = _StartTime,
                                        .EndTime = _EndTime,
                                        .IsSingleDay = _EndTime - _StartTime <= TimeSpan.FromDays(1),
                                        .IsTimeSpecified = e.<all_day>.Value.Equals("0"),
                                        .Artists = (From a In e.<performers>.<performer>
                                                    Select New Artist() With
                                                           {
                                                               .Id = a.<id>.Value,
                                                               .Name = a.<name>.Value,
                                                               .Uri = New Uri(a.<url>.Value)
                                                           }).ToList(),
                                         .Description = e.<description>.Value,
                                         .Image = If(Not [String].IsNullOrEmpty(e.<image>.<url>.Value) AndAlso Not e.<image>.<url>.Value.EndsWith("gif"), New Uri(e.<image>.<url>.Value), Nothing),
                                         .Uri = New Uri(e.<url>.Value)
                                    }

                Return New EventsAvailable(events.ToList(), EventGroup.Venue, Id)
            Catch ex As Exception
                Return New EventsAvailable(New List(Of Model.Event)(), EventGroup.Venue, Id)
            End Try
        End Function

        ''' <summary>
        ''' Parse the given search response XML based on the given type.
        ''' </summary>
        ''' <param name="Type">The type of the search results.</param>
        ''' <param name="Xml">The search response XML.</param>
        ''' <returns>A list of ISearchable.</returns>
        Private Function ParseSearchResults([Type] As [Type], Xml As XElement) As List(Of ISearchable)
            If [Type].[Equals](GetType(Artist)) Then
                Dim Artists = From e In Xml.<performers>.<performer>
                              Select TryCast(New Artist() With
                                             {
                                                 .Id = e.<id>.Value,
                                                 .Name = e.<name>.Value,
                                                 .Image = If(Not String.IsNullOrEmpty(e.<image>...<url>.Value), New Uri(e.<image>...<url>.Value), Nothing),
                                                 .Uri = New Uri(e.<url>.Value)
                                             }, ISearchable)
                Return Enumerable.ToList(Of ISearchable)(Artists)
            ElseIf [Type].[Equals](GetType(Venue)) Then
                Dim Venues = From e In Xml.<venues>.<venue>
                            Select TryCast(New Venue() With
                                            {
                                                .Id = e.@id,
                                                .Name = e.<venue_name>.Value,
                                                .Image = If(Not String.IsNullOrEmpty(e.<image>.<url>.Value) AndAlso Not e.<image>.<url>.Value.EndsWith("gif"), New Uri(e.<image>.<url>.Value), Nothing),
                                                .Uri = New Uri(e.<url>.Value)
                                            }, ISearchable)
                Return Enumerable.ToList(Of ISearchable)(Venues)
            Else
                Return New List(Of ISearchable)()
            End If
        End Function

    End Class

End Namespace

File: externals/Async-CTP/samples/CS/Windows/Phone/Concert/Finder/Concert/Finder/Providers/EventfulProvider.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Net;
using System.Threading;
using System.Threading.Tasks;
using System.Xml.Linq;
using ConcertFinder.Model;
using ConcertFinder.Resources;

namespace ConcertFinder.Providers
{
    /// <summary>Eventful event data provider.</summary>
    public class EventfulProvider : IProvider
    {
        /// <summary>Search for artists and venues based on the given query.</summary>
        /// <param name="query">The search query.</param>
        /// <param name="cancellationToken">The cancellation token.</param>
        public Task<SearchResults[]> SearchAsync(string query, CancellationToken cancellationToken)
        {
            var artistsUri = new Uri(
                String.Format("http://api.eventful.com/rest/performers/search?keywords={0}&page_size=100&app_key={1}",
                query, Configuration.EventfulAppID));

            var venuesUri = new Uri(
                String.Format("http://api.eventful.com/rest/venues/search?keywords={0}&page_size=100&app_key={1}",
                query, Configuration.EventfulAppID));

            var artists = DownloadAndParse(artistsUri, query, typeof(Model.Artist), cancellationToken);
            var venues = DownloadAndParse(venuesUri, query, typeof(Model.Venue), cancellationToken);
            return TaskEx.WhenAll(artists, venues);
        }

        /// <summary>Downloads and parses the provided URI.</summary>
        /// <param name="uri">The URI to download for results.</param>
        /// <param name="query">The search query.</param>
        /// <param name="searchType">The search model type.</param>
        /// <param name="cancellationToken">The cancellation token.</param>
        private async Task<SearchResults> DownloadAndParse(
            Uri uri, string query, Type searchType, CancellationToken cancellationToken)
        {
            try
            {
                var content = await new WebClient().DownloadStringTaskAsync(uri, cancellationToken);
                var results = ParseSearchResults(searchType, XElement.Parse(content));
                return new SearchResults(results, searchType, query);
            }
            catch
            {
                return new SearchResults(new List<ISearchable>(), searchType, query);
            }
        }

        /// <summary>Returns a list of events for the artist with the given id.</summary>
        /// <param name="id">The artist id.</param>
        public async Task<EventsAvailable> GetArtistEventsAsync(string id)
        {
            var uri = new Uri(
               String.Format("http://api.eventful.com/rest/events/search?keywords=spid:{0}&page_size=100&category=music&sort_order=date&app_key={1}",
               id, Configuration.EventfulAppID));

            try
            {
                var parsed = XElement.Parse(await new WebClient().DownloadStringTaskAsync(uri));

                var events = from e in parsed.Element("events").Elements("event")
                             let _City = e.Element("city_name").Value
                             let _Region = e.Element("region_name").Value
                             let _Country = e.Element("country_name").Value
                             let _StartTime = !String.IsNullOrEmpty(e.Element("start_time").Value) ?
                                    (DateTime?)Convert.ToDateTime(e.Element("start_time").Value) : null
                             let _EndTime = !String.IsNullOrEmpty(e.Element("stop_time").Value) ?
                                    (DateTime?)Convert.ToDateTime(e.Element("stop_time").Value) : null
                             where !(_StartTime.HasValue && _EndTime.HasValue &&
                                     _StartTime.Value.AddDays(1).CompareTo(_EndTime.Value) < 0)
                             select new Model.Event()
                             {
                                 Id = e.Attribute("id").Value,
                                 Name = e.Element("title").Value,
                                 Location = String.Format("{0}, {1}", _Country.ToLower().Equals("united states") ? _Region : _City, _Country),
                                 Venue = new Venue()
                                 {
                                     Id = e.Element("venue_id").Value,
                                     Name = e.Element("venue_name").Value,
                                     Address = new Address()
                                     {
                                         Street = e.Element("venue_address").Value,
                                         City = e.Element("city_name").Value,
                                         Region = e.Element("region_name").Value,
                                         Country = e.Element("country_name").Value,
                                         Latitude = !String.IsNullOrEmpty(e.Element("latitude").Value) ?
                                            Convert.ToDouble(e.Element("latitude").Value, CultureInfo.InvariantCulture.NumberFormat) : (double?)null,
                                         Longitude = !String.IsNullOrEmpty(e.Element("longitude").Value) ?
                                            Convert.ToDouble(e.Element("longitude").Value, CultureInfo.InvariantCulture.NumberFormat) : (double?)null
                                     },
                                     Uri = e.Element("venue_url") != null &&
                                         !String.IsNullOrEmpty(e.Element("venue_url").Value) ?
                                         new Uri(e.Element("venue_url").Value) : null
                                 },
                                 StartTime = _StartTime,
                                 EndTime = _EndTime,
                                 IsSingleDay = !_EndTime.HasValue || _StartTime.HasValue &&
                                 _EndTime.Value - _StartTime.Value <= TimeSpan.FromDays(1),
                                 IsTimeSpecified = e.Element("all_day").Value.Equals("0"),
                                 Artists = (from a in e.Element("performers").Elements("performer")
                                            select new Artist()
                                            {
                                                Id = a.Element("id").Value,
                                                Name = a.Element("name").Value,
                                                Uri = new Uri(a.Element("url").Value)
                                            }).ToList<Artist>(),
                                 Description = e.Element("description").Value,
                                 Image = e.Element("image").Element("url") != null &&
                                    !String.IsNullOrEmpty(e.Element("image").Element("url").Value) &&
                                    !e.Element("image").Element("url").Value.EndsWith("gif") ?
                                    new Uri(e.Element("image").Element("url").Value) :
                                    null,
                                 Uri = new Uri(e.Element("url").Value)
                             };

                return new EventsAvailable(events.ToList(), EventGroup.Artist, id);
            }
            catch 
            {
                return new EventsAvailable(new List<Model.Event>(), EventGroup.Artist, id);
            }
        }

        /// <summary>
        /// Returns a list of events for the venue with the given id.
        /// </summary>
        /// <param name="id">The venue id.</param>
        public async Task<EventsAvailable> GetVenueEventsAsync(string id)
        {
            var uri = new Uri(
               String.Format("http://api.eventful.com/rest/events/search?location={0}&page_size=100&category=music&sort_order=date&app_key={1}",
               id, Configuration.EventfulAppID));

            try
            {
                var parsed = XElement.Parse(await new WebClient().DownloadStringTaskAsync(uri));

                var events = from e in parsed.Element("events").Elements("event")
                             let _City = e.Element("city_name").Value
                             let _Region = e.Element("region_name").Value
                             let _Country = e.Element("country_name").Value
                             let _StartTime = !String.IsNullOrEmpty(e.Element("start_time").Value) ?
                                    (DateTime?)Convert.ToDateTime(e.Element("start_time").Value) :
                                    null
                             let _EndTime = !String.IsNullOrEmpty(e.Element("stop_time").Value) ?
                                    (DateTime?)Convert.ToDateTime(e.Element("stop_time").Value) :
                                    null
                             where !(_StartTime.HasValue && _EndTime.HasValue &&
                                     _StartTime.Value.AddDays(1).CompareTo(_EndTime.Value) < 0)
                             select new Model.Event()
                             {
                                 Id = e.Attribute("id").Value,
                                 Name = e.Element("title").Value,
                                 Location = e.Element("venue_name").Value,
                                 Venue = new Venue()
                                 {
                                     Id = e.Element("venue_id").Value,
                                     Name = e.Element("venue_name").Value,
                                     Address = new Address()
                                     {
                                         Street = e.Element("venue_address").Value,
                                         City = e.Element("city_name").Value,
                                         Region = e.Element("region_name").Value,
                                         Country = e.Element("country_name").Value,
                                         Latitude = !String.IsNullOrEmpty(e.Element("latitude").Value) ?
                                            Convert.ToDouble(e.Element("latitude").Value, CultureInfo.InvariantCulture.NumberFormat) :
                                            (double?)null,
                                         Longitude = !String.IsNullOrEmpty(e.Element("longitude").Value) ?
                                            Convert.ToDouble(e.Element("longitude").Value, CultureInfo.InvariantCulture.NumberFormat) :
                                            (double?)null
                                     },
                                     Uri = e.Element("venue_url") != null &&
                                         !String.IsNullOrEmpty(e.Element("venue_url").Value) ?
                                         new Uri(e.Element("venue_url").Value) :
                                         null
                                 },
                                 StartTime = _StartTime,
                                 EndTime = _EndTime,
                                 IsSingleDay = !_EndTime.HasValue || _StartTime.HasValue &&
                                 _EndTime.Value - _StartTime.Value <= TimeSpan.FromDays(1),
                                 IsTimeSpecified = e.Element("all_day").Value.Equals("0"),
                                 Artists = (from a in e.Element("performers").Elements("performer")
                                            select new Artist()
                                            {
                                                Id = a.Element("id").Value,
                                                Name = a.Element("name").Value,
                                                Uri = new Uri(a.Element("url").Value)
                                            }).ToList<Artist>(),
                                 Description = e.Element("description").Value,
                                 Image = e.Element("image").Element("url") != null &&
                                    !String.IsNullOrEmpty(e.Element("image").Element("url").Value) &&
                                    !e.Element("image").Element("url").Value.EndsWith("gif") ?
                                    new Uri(e.Element("image").Element("url").Value) :
                                    null,
                                 Uri = new Uri(e.Element("url").Value)
                             };

                return new EventsAvailable(events.ToList<Model.Event>(), EventGroup.Venue, id);
            }
            catch 
            {
                return new EventsAvailable(new List<Model.Event>(), EventGroup.Venue, id);
            }
        }

        /// <summary>
        /// Parse the given search response XML based on the given type.
        /// </summary>
        /// <param name="Type">The type of the search results.</param>
        /// <param name="Xml">The search response XML.</param>
        /// <returns>A list of ISearchable.</returns>
        private List<ISearchable> ParseSearchResults(Type Type, XElement Xml)
        {
            if (Type.Equals(typeof(Artist)))
            {
                var Artists = from e in Xml.Element("performers").Elements("performer")
                              select new Artist()
                              {
                                  Id = e.Element("id").Value,
                                  Name = e.Element("name").Value,
                                  Image = e.Element("image").Descendants("url").Count() > 0 &&
                                     !String.IsNullOrEmpty(e.Element("image").Descendants("url").First().Value) ?
                                     new Uri(e.Element("image").Descendants("url").First().Value) :
                                     null,
                                  Uri = new Uri(e.Element("url").Value)
                              } as ISearchable;

                return Artists.ToList<ISearchable>();
            }
            else if (Type.Equals(typeof(Venue)))
            {
                var Venues = from e in Xml.Element("venues").Elements("venue")
                             select new Venue()
                             {
                                 Id = e.Attribute("id").Value,
                                 Name = e.Element("venue_name").Value,
                                 Image = e.Element("image").Element("url") != null &&
                                    !String.IsNullOrEmpty(e.Element("image").Element("url").Value) &&
                                    !e.Element("image").Element("url").Value.EndsWith("gif") ?
                                    new Uri(e.Element("image").Element("url").Value) :
                                    null,
                                 Uri = new Uri(e.Element("url").Value)
                             } as ISearchable;

                return Venues.ToList<ISearchable>();
            }
            else
            {
                return new List<ISearchable>();
            }
        }
    }
}